### ISSUE TITLE
Replicate identity errors on a table with no relations

### ISSUE THREAD

**Issue by programbo** [prisma/prisma#9434]
### Bug description

Using NextAuth.js with the Prisma adapter, my sign in process is interrupted by a SQL error. The error is caused by attempting to delete a row from a table without a replica identity. This is unexpected as the table has no relations (See `VerificationToken` model in the schema below.

Changing the `token` field from `@unique` to `@id` eliminates the error however it is unclear why a primary key prevents an error that should not occur in the first place.

This error appears to be isolated as it does not occur in the NextAuth.js adapter tests suite. cc @balazsorban44.

### How to reproduce

1. Generate Magic Link email from NextAuth.js (adds a row to the `VerificationToken` table)
2. Click the "Sign in" link in the email (uses and then attempts to delete the row from `VerificationToken`
3. PrismaAdapter throws an error
4. See error log

### Expected behavior

1. Row should be deleted, and
2. Table without relations should not have replica identity errors.

#### Schema

```prisma
generator client {
  provider        = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Upload {
  id        String   @id @default(uuid())
  userId    String
  center    String
  data      Json
  createdAt DateTime @default(now())
  user      User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}

model Account {
  id                 String  @id @default(cuid())
  userId             String
  type               String
  provider           String
  providerAccountId  String
  refresh_token      String?
  access_token       String?
  expires_at         Int?
  token_type         String?
  scope              String?
  id_token           String?
  session_state      String?
  oauth_token_secret String?
  oauth_token        String?
  user               User    @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([provider, providerAccountId])
}

model Session {
  id           String   @id @default(cuid())
  sessionToken String   @unique
  userId       String
  expires      DateTime
  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}

model User {
  id            String    @id @default(cuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  image         String?
  accounts      Account[]
  sessions      Session[]
  uploads       Upload[]
}

model VerificationToken {
  identifier String
  token      String   @unique
  expires    DateTime

  @@unique([identifier, token])
}
```

#### Client query in [NextAuth.js Prisma adapter v0.5.2-next.15](https://github.com/nextauthjs/adapters/blob/bc84a1cbfe55327f6d94de395aadf96cc0beef40/packages/prisma/src/index.ts#L36-L46)

```ts
async useVerificationToken(identifier_token) {
  try {
    return await p.verificationToken.delete({ where: { identifier_token } })
  } catch (error) {
    // If token already used/deleted, just return null
    // https://www.prisma.io/docs/reference/api-reference/error-reference#p2025
    if ((error as Prisma.PrismaClientKnownRequestError).code === "P2025")
      return null
    throw error
  }
},
```

#### Error log

```
2021-09-25T15:41:26.075Z prisma:client clientVersion: 3.1.1
2021-09-25T15:41:26.080Z prisma:client:libraryEngine internalSetup
2021-09-25T15:41:26.105Z prisma:client:libraryEngine Search for Query Engine Library in /Users/john.lombardo/Projects/droploader/node_modules/.prisma/client
2021-09-25T15:41:26.105Z prisma:client:libraryEngine loadEngine using /Users/john.lombardo/Projects/droploader/node_modules/.prisma/client/libquery_engine-darwin.dylib.node
2021-09-25T15:41:26.138Z prisma:client Prisma Client call:
2021-09-25T15:41:26.140Z prisma:client prisma.verificationToken.delete({
  where: {
    identifier_token: {
      identifier: 'hello@johnlombardo.dev',
      token: 'c8e6a63d94197be345692246dd72b3499af5744f42225b9cd0e15bad4e07c28c'
    }
  }
})
2021-09-25T15:41:26.140Z prisma:client Generated request:
2021-09-25T15:41:26.140Z prisma:client mutation {
  deleteOneVerificationToken(where: {
    identifier_token: {
      identifier: "hello@johnlombardo.dev"
      token: "c8e6a63d94197be345692246dd72b3499af5744f42225b9cd0e15bad4e07c28c"
    }
  }) {
    identifier
    token
    expires
  }
}

2021-09-25T15:41:26.142Z prisma:client:libraryEngine sending request, this.libraryStarted: false
2021-09-25T15:41:26.142Z prisma:client:libraryEngine library starting
2021-09-25T15:41:26.171Z prisma:client:libraryEngine library started
2021-09-25T15:41:26.197Z prisma:client:libraryEngine graphQLToJSError
2021-09-25T15:41:26.197Z prisma:client:fetcher Error: Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState("55000"), message: "cannot delete from table \"VerificationToken\" because it does not have a replica identity and publishes deletes", detail: None, hint: Some("To enable deleting from the table, set REPLICA IDENTITY using ALTER TABLE."), position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("execReplication.c"), line: Some(583), routine: Some("CheckCmdReplicaIdentity") }) }) })
    at LibraryEngine.prismaGraphQLToJSError (/Users/john.lombardo/Projects/droploader/node_modules/@prisma/client/runtime/index.js:25140:14)
    at LibraryEngine.request (/Users/john.lombardo/Projects/droploader/node_modules/@prisma/client/runtime/index.js:25155:24)
    at async cb (/Users/john.lombardo/Projects/droploader/node_modules/@prisma/client/runtime/index.js:36922:26)
    at async useVerificationToken (/Users/john.lombardo/Projects/droploader/node_modules/@next-auth/prisma-adapter/dist/index.js:37:24)
    at async acc.<computed> (webpack-internal:///./node_modules/next-auth/server/errors.js:92:16)
    at async Object.callback (webpack-internal:///./node_modules/next-auth/server/routes/callback.js:178:22)
    at async NextAuthHandler (webpack-internal:///./node_modules/next-auth/server/index.js:171:18)
    at async eval (webpack-internal:///./node_modules/next-auth/server/index.js:251:32)
    at async Object.apiResolver (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/api-utils.js:101:9)
    at async DevServer.handleApiRequest (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/next-server.js:770:9)
[next-auth][error][adapter_error_useVerificationToken] 
https://next-auth.js.org/errors#adapter_error_useverificationtoken 
Invalid `prisma.verificationToken.delete()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState("55000"), message: "cannot delete from table \"VerificationToken\" because it does not have a replica identity and publishes deletes", detail: None, hint: Some("To enable deleting from the table, set REPLICA IDENTITY using ALTER TABLE."), position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("execReplication.c"), line: Some(583), routine: Some("CheckCmdReplicaIdentity") }) }) }) {
  message: '\n' +
    'Invalid `prisma.verificationToken.delete()` invocation:\n' +
    '\n' +
    '\n' +
    '  Error occurred during query execution:\n' +
    'ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState("55000"), message: "cannot delete from table \\"VerificationToken\\" because it does not have a replica identity and publishes deletes", detail: None, hint: Some("To enable deleting from the table, set REPLICA IDENTITY using ALTER TABLE."), position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("execReplication.c"), line: Some(583), routine: Some("CheckCmdReplicaIdentity") }) }) })',
  stack: 'Error: \n' +
    'Invalid `prisma.verificationToken.delete()` invocation:\n' +
    '\n' +
    '\n' +
    '  Error occurred during query execution:\n' +
    'ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState("55000"), message: "cannot delete from table \\"VerificationToken\\" because it does not have a replica identity and publishes deletes", detail: None, hint: Some("To enable deleting from the table, set REPLICA IDENTITY using ALTER TABLE."), position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("execReplication.c"), line: Some(583), routine: Some("CheckCmdReplicaIdentity") }) }) })\n' +
    '    at cb (/Users/john.lombardo/Projects/droploader/node_modules/@prisma/client/runtime/index.js:36956:17)\n' +
    '    at async useVerificationToken (/Users/john.lombardo/Projects/droploader/node_modules/@next-auth/prisma-adapter/dist/index.js:37:24)\n' +
    '    at async acc.<computed> (webpack-internal:///./node_modules/next-auth/server/errors.js:92:16)\n' +
    '    at async Object.callback (webpack-internal:///./node_modules/next-auth/server/routes/callback.js:178:22)\n' +
    '    at async NextAuthHandler (webpack-internal:///./node_modules/next-auth/server/index.js:171:18)\n' +
    '    at async eval (webpack-internal:///./node_modules/next-auth/server/index.js:251:32)\n' +
    '    at async Object.apiResolver (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/api-utils.js:101:9)\n' +
    '    at async DevServer.handleApiRequest (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/next-server.js:770:9)\n' +
    '    at async Object.fn (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/next-server.js:661:37)\n' +
    '    at async Router.execute (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/router.js:205:32)',
  name: 'Error'
}
[next-auth][error][CALLBACK_EMAIL_ERROR] 
https://next-auth.js.org/errors#callback_email_error 
Invalid `prisma.verificationToken.delete()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState("55000"), message: "cannot delete from table \"VerificationToken\" because it does not have a replica identity and publishes deletes", detail: None, hint: Some("To enable deleting from the table, set REPLICA IDENTITY using ALTER TABLE."), position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("execReplication.c"), line: Some(583), routine: Some("CheckCmdReplicaIdentity") }) }) }) Error: 
Invalid `prisma.verificationToken.delete()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState("55000"), message: "cannot delete from table \"VerificationToken\" because it does not have a replica identity and publishes deletes", detail: None, hint: Some("To enable deleting from the table, set REPLICA IDENTITY using ALTER TABLE."), position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("execReplication.c"), line: Some(583), routine: Some("CheckCmdReplicaIdentity") }) }) })
    at cb (/Users/john.lombardo/Projects/droploader/node_modules/@prisma/client/runtime/index.js:36956:17)
    at async useVerificationToken (/Users/john.lombardo/Projects/droploader/node_modules/@next-auth/prisma-adapter/dist/index.js:37:24)
    at async acc.<computed> (webpack-internal:///./node_modules/next-auth/server/errors.js:92:16)
    at async Object.callback (webpack-internal:///./node_modules/next-auth/server/routes/callback.js:178:22)
    at async NextAuthHandler (webpack-internal:///./node_modules/next-auth/server/index.js:171:18)
    at async eval (webpack-internal:///./node_modules/next-auth/server/index.js:251:32)
    at async Object.apiResolver (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/api-utils.js:101:9)
    at async DevServer.handleApiRequest (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/next-server.js:770:9)
    at async Object.fn (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/next-server.js:661:37)
    at async Router.execute (/Users/john.lombardo/Projects/droploader/node_modules/next/dist/server/router.js:205:32) {
  name: 'UseVerificationTokenError'
}
```

### Environment & setup

- OS: MacOS 11.6
- Database: PostgreSQL 12.3 (Debian 12.3-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit (included in Supabase local dev container)
- Node.js version: v14.17.6

### Prisma Version

```
prisma                  : 3.1.1
@prisma/client          : 3.1.1
Current platform        : darwin
Query Engine (Node-API) : libquery-engine c22652b7e418506fab23052d569b85d3aec4883f (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli c22652b7e418506fab23052d569b85d3aec4883f (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core c22652b7e418506fab23052d569b85d3aec4883f (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt c22652b7e418506fab23052d569b85d3aec4883f (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash    : c22652b7e418506fab23052d569b85d3aec4883f
Studio                  : 0.423.0
```

**Comment by pantharshit00** [comment]
@programbo Can you please a reproduction inside of git repo that we can easily use? It is hard to replicate your setup just with the description. 

Also, please make sure the bug doesn't exist in nextauth itself, otherwise we will need to transfer this there.

**Comment by balazsorban44** [comment]
NextAuth.js maintainer here. I advised checking in here from https://github.com/nextauthjs/docs/issues/46

We run tests with Prisma, and those don't reproduce this problem.

A reproduction indeed would be most helpful.

**Comment by programbo** [comment]
Should be able to do that this weekend. I'll also test it against a hosted PostgreSQL database to to ensure it's not something specific to my local DB.

**Comment by pantharshit00** [comment]
I am going to close this as its been 3 months now and there is no activity here. 

Please reply if you want us to take another look.

**Comment by sdgluck** [comment]
I just encountered this issue myself. For us the issue was due to Prisma relation tables not having a primary key and so under the [default replica identity behaviour](https://www.postgresql.org/docs/current/logical-replication-publication.html) are not able to be replicated without further configuration. 

Prisma relation tables use a unique index on the two columns A and B and so theoretically the `REPLICA IDENTITY FULL` should be able to perform without issue as this constraint will prevent duplicate records which would interfere with replication.

I see two options for a resolution:
1. include `ALTER TABLE x REPLICA IDENTITY FULL` in the SQL for creation of these tables in migrations
2. add a primary key on relation tables

My preference would be 2 as it means future join table schemas will not diverge from the current schema.
